blocks/fast_reader_test.go

```diff
@@ -36,33 +47,57 @@ func TestFastRead(t *testing.T) {
 	}
 
 	for i, test := range tests {
-		buffer := new(bytes.Buffer)
-		w := NewWriter(buffer, test.blockSize)
-
-		w.Write([]byte(test.input))
-		w.Flush()
-
-		r := NewFastReader(context.Background(), bytes.NewReader(buffer.Bytes()), test.blockSize)
-
-		var err error
-		var length int
-		result := make([]byte, 0)
-		bytes := make([]byte, test.readSize)
-		n := 1
-
-		for n > 0 && err == nil {
-			n, err = r.Read(bytes)
-			result = append(result, bytes[:n]...)
-			length += n
-		}
-
-		if !reflect.DeepEqual(result, []byte(test.input)) {
-			t.Errorf("Wrong bytes for Case %d:\n want: %x\n  got: %x", i, test.input, result)
-		}
-
-		if length != len(test.input) || err != io.EOF {
-			t.Errorf("Wrong return for Case %d: want: %d,EOF got: %d,%v", i, len(test.input), length, err)
-		}
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			buffer := new(bytes.Buffer)
+			w := NewWriter(buffer, test.blockSize)
+
+			w.Write([]byte(test.input))
+			w.Flush()
+
+			// r := NewFastReader(context.Background(), bytes.NewReader(buffer.Bytes()), test.blockSize)
+			f := bytes.NewReader(buffer.Bytes())
+			fr := fileReaderPool.Get().(*bufio.Reader)
+			fr.Reset(f)
+
+			// defer func() {
+			// 	fr.Reset(nil)
+			// 	fileReaderPool.Put(fr)
+			// }()
+
+			ctx, cancel := context.WithCancel(context.Background())
+			r := NewFastReader(ctx, fr, test.blockSize)
+
+			defer func() {
+				cancel()
+				r.Close()
+				fr.Reset(nil)
+				fileReaderPool.Put(fr)
+			}()
```

Comment:
This updated test would crash under `go test -race -v ./blocks/...`. I tried adding the cancellation like you did in https://github.com/customerio/services/pull/11747 , but it still crashed.

I think it still crashes, because it's still racy. There is no guarantee with cancellation that the context cancellation is acted on in `readAhead` go routine _before_ the `Reset` is called on the `bufio.Reader`.

This is the kind of race I saw before I added the `sync.WaitGroup` and the `Close` function.

```
rdawe@Richard-Dawe esdb % go test -race -v ./blocks/...
=== RUN   TestFastRead
=== RUN   TestFastRead/0
=== RUN   TestFastRead/1
==================
WARNING: DATA RACE
Read at 0x00c000038268 by goroutine 8:
  bufio.(*Reader).Read()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/bufio/bufio.go:220 +0x114
  github.com/customerio/esdb/blocks.(*FastReader).readAhead()
      /Users/rdawe/code/esdb/blocks/fast_reader.go:203 +0x1c8
  github.com/customerio/esdb/blocks.NewFastReader.gowrap1()
      /Users/rdawe/code/esdb/blocks/fast_reader.go:81 +0x4c

Previous write at 0x00c000038268 by goroutine 7:
  bufio.(*Reader).reset()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/bufio/bufio.go:85 +0x104
  bufio.(*Reader).Reset()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/bufio/bufio.go:81 +0xb8
  github.com/customerio/esdb/blocks.TestFastRead.func1.1()
      /Users/rdawe/code/esdb/blocks/fast_reader_test.go:70 +0x48
  runtime.deferreturn()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/runtime/panic.go:602 +0x5c
  testing.tRunner()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/testing/testing.go:1742 +0x40

Goroutine 8 (running) created at:
  github.com/customerio/esdb/blocks.NewFastReader()
      /Users/rdawe/code/esdb/blocks/fast_reader.go:81 +0x3f4
  github.com/customerio/esdb/blocks.TestFastRead.func1()
      /Users/rdawe/code/esdb/blocks/fast_reader_test.go:73 +0x3cc
  testing.tRunner()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/testing/testing.go:1742 +0x40

Goroutine 7 (finished) created at:
  testing.(*T).Run()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/testing/testing.go:1742 +0x5e4
  github.com/customerio/esdb/blocks.TestFastRead()
      /Users/rdawe/code/esdb/blocks/fast_reader_test.go:50 +0x19c
  testing.tRunner()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /Users/rdawe/.local/share/mise/installs/go/1.22.7/src/testing/testing.go:1742 +0x40
==================
```